src/components/compositions/notifications/index.ts

```diff
@@ -0,0 +1,135 @@
+import { computed, ref, watch } from "@vue/composition-api";
+import { web3 } from "@/api/web3";
+
+export enum ENotificationStatus {
+  success = "success",
+  info = "info",
+  warning = "warning",
+  error = "error",
+  pending = "pending"
+}
+
+export interface INotificationView {
+  id: number;
+  status: ENotificationStatus;
+  title: string;
+  description: string;
+  txHash?: string;
+  timestamp: Date;
+  showSeconds?: number;
+  alertOnly?: boolean;
+}
+
+export const history = ref<INotificationView[]>([]);
+
+export const alertQueue = computed(() =>
+  history.value.filter(n => n.showSeconds).reverse()
+);
+
+export const pendingQueue = computed(() =>
+  history.value.filter(n => n.status === ENotificationStatus.pending)
+);
+
+export const addNotification = (
+  title: string,
+  description = "",
+  status = ENotificationStatus.info,
+  txHash?: string,
+  showSeconds = 8,
+  alertOnly?: boolean
+) => {
+  const notification: INotificationView = {
+    id: Date.now(),
+    status,
+    title,
+    description,
+    txHash,
+    showSeconds,
+    alertOnly,
+    timestamp: new Date()
+  };
+
+  if (txHash) {
+    notification.status = ENotificationStatus.pending;
+  }
+
+  history.value.unshift(notification);
+};
+
+export const removeNotification = (id: number) => {
+  const index = history.value
+    .map(n => {
+      return n.id;
+    })
+    .indexOf(id);
+
+  if (index >= 0) history.value.splice(index, 1);
+  else console.error("Error: Remove Notification - ID not found.");
+};
+
+export const hideAlert = (id: number) => {
+  const index = history.value
+    .map(n => {
+      return n.id;
+    })
+    .indexOf(id);
+
+  if (index >= 0) {
+    if (history.value[index].alertOnly) history.value.splice(index, 1);
+    else history.value[index].showSeconds = undefined;
+  } else console.error("Error: Hide Alert - ID not found.");
+};
+
+export const clearAllNotifications = () => {
+  history.value = [];
+};
+
+export const getTxStatus = async (hash: string) => {
+  try {
+    const tx = await web3.eth.getTransactionReceipt(hash);
+    if (tx) return tx.status;
+    else return null;
+  } catch (e) {
+    console.error(" - web3 failed: getTransactionReceipt - ", e);
+    throw e;
+  }
+};
+
+export const updatePendingTx = async (notification: INotificationView) => {
+  try {
+    const status = await getTxStatus(notification.txHash!);
+    if (status !== null) {
+      // set status
+      if (status) notification.status = ENotificationStatus.success;
+      else notification.status = ENotificationStatus.error;
+
+      // find history index
+      const index = history.value
+        .map(n => {
+          return n.txHash;
+        })
+        .indexOf(notification.txHash);
```

Comment:
Can we use the ```updateArray``` function here? Or do we have some performance concerns, not entirely sure how long this thing might end up being. Is there a limit as to how much we're gonna store in localStorage?

Reply:
I was thinking max 100 notifications.

`updateArray` probably works fine. I was just trying to keep the dependencies at a minimum so that this composition can be reused in another project if needed.